### DISCUSSION TITLE
New security rules

### DISCUSSION THREAD

**Discussion by icy17** [openssl/openssl#24627] [Q&A]
Hi, I've observed that some security rules are absent from the documentation, such as for API `EVP_DigestInit_ex`, which requires that parameter 1 cannot be NULL. Omitting these rules could result in overlooked necessary checks when using the API. Should I report these types of rules to your team?

**Comment by nhorman** [comment]
You are welcome to report whatever concerns you like.  We always welcome feedback, bug reports and feature requests (and their associated PRs)

However, reports of this nature (in which a vague problem is stated, but no specific action is requested are difficult to triage and plan.  Please open a new issue for whatever subject you have in mind, provide as much detail as possible, and if you are able, please feel free to open a pull request to suggest a corrective measure.  Your participation is greatly appreciated!

**Comment by icy17** [comment]
Hi, I've observed that some security rules are absent from the documentation.

For the security rule "parameters must not be NULL," its omission in the documentation can lead to significant security issues.

Firstly, while best programming practices dictate that developers should avoid passing NULL parameters to an API, this guideline may not be obvious to less experienced developers, increasing the risk of overlooking necessary security checks.

Secondly, many APIs receive parameters from other functions or APIs, which might not be adequately checked for NULL, posing a risk even for experienced developers.

Therefore, it is crucial to explicitly mention in the documentation that NULL checks must be conducted on specified parameters before an API is used. Including these security rules will help developers identify potential security issues more effectively during testing.

I will provide a list of APIs requiring checks along with their corresponding parameters：
| API | parameter index (begin with 1) | 
|:----------:|:----------:|
| MD5_Init | 1 |
| SHA512_Update | 2 |
| SHA256_Init | 1 |
| EVP_DigestSignUpdate | 1 |
|BN_bin2bn | 1 |
| SHA512_Final | 1 |
| SHA512_Init | 1 |
| X509_get_subject_name | 1 |
| BIO_new | 1 |
| DES_ecb_encrypt | 1 |
| DES_ecb_encrypt | 2 |
| BN_bn2bin | 1 |
| BN_bn2bin | 2 |
| ERR_error_string_n | 2 |
| OPENSSL_sk_delete_ptr | 1 |
| OPENSSL_sk_set_cmp_func | 1 |
| SSL_ctrl | 1 |
| SSL_CTX_set_cipher_list | 1 |
| SSL_shutdown | 1 |
| SSL_CTX_set_verify | 1 |
| SHA384_Update | 1 |
| EVP_DigestInit_ex | 1 |
| SSL_CTX_load_verify_locations | 1 |
| EVP_DigestFinal_ex | 1 |
| EVP_DigestFinal_ex | 2 |
| SSL_want | 1 |
| ASN1_STRING_get0_data | 1 |
| ASN1_STRING_length | 1 |
| SSL_CTX_use_PrivateKey_file| 1 |
| OSSL_PROVIDER_load | 2 |
| SSL_CTX_get_cert_store | 1 |
| SSL_set_ex_data | 1 |
| SSL_get_verify_result | 1 |
| EVP_DigestInit | 1 |
| SSL_connect | 1 |
| SSL_CTX_use_PrivateKey | 1 |
| DES_ede3_cbc_encrypt | 1 |
| DES_ede3_cbc_encrypt | 7 |
| SHA384_Init | 1 |
| RAND_bytes | 1 |
| SSL_CTX_set_default_verify_paths | 1 |
| X509_STORE_CTX_get_error | 1 |
| EVP_CIPHER_CTX_set_padding | 1 |
| SHA224_Final | 1 |
| EVP_EncryptInit_ex | 1 |
| EVP_PKEY_new_mac_key | 3 |
| SHA224_Init | 1 |
| SSL_set_connect_state | 1 |
| EVP_EncryptUpdate | 1 |
| EVP_EncryptUpdate | 4 |
| X509_get0_notAfter | 1 |
| X509_get0_notBefore | 1 |
| SSL_is_init_finished | 1 |
| SSL_set_info_callback | 1 |
| X509_STORE_CTX_get_error_depth | 1 |
| X509_STORE_set_flags | 1 |
| SSL_get_current_cipher | 1 |
| PEM_read_bio_PrivateKey | 1 |
| X509_STORE_CTX_set_error | 1 |
| EVP_CIPHER_get0_name | 1 |
| X509_STORE_CTX_get_ex_data | 1 |
| EVP_DecryptInit_ex | 1 |
| SSL_CTX_set_options | 1 |

**Comment by icy17** [comment]
There are other types of security rules that are missing in the documentation. I believe incorporating these rules could help developers learn how to correctly use APIs and reduce the risk of security issues.

I will provide a table describing the missing security rules:
| API | parameter index (begin with 1) | Security rules |
|:----------:|:----------:|:----------:|
| BN_free | 1 | parameter must not be used later (without re-initialize) |
| X509_free | 1 | parameter must not be used later (without re-initialize) |
| BIO_free | 1 | parameter must not be used later (without re-initialize) |
| SSL_CTX_free | 1 | parameter must not be used later (without re-initialize) |
| SSL_free | 1 | parameter must not be used later (without re-initialize) |
| EVP_MD_free | 1 | parameter must not be used later (without re-initialize) |
| SSL_CTX_free | 1 | parameter must not be used later (without re-initialize) |
| EVP_MD_CTX_free | 1 | parameter must not be used later (without re-initialize) |
|  OSSL_LIB_CTX_free | 1 | parameter must not be used later (without re-initialize) |
|  BN_CTX_free | 1 | parameter must not be used later (without re-initialize) |
|  X509_STORE_CTX_free | 1 | parameter must not be used later (without re-initialize) |
|  EVP_PKEY_free | 1 | parameter must not be used later (without re-initialize) |
|  CRYPTO_free | 1 | parameter must not be used later (without re-initialize) |
| BIO_free | 1 | parameter must be initialized using xxx (other API) |
| EVP_PKEY_free | 1 | parameter must be initialized using xxx (other API) |
| DES_set_key_unchecked | 2 | parameter must be initialized using xxx (other API) |
| BIO_ctrl | 1 | parameter must be initialized using xxx (other API) |
| SHA512_Final | 2 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_value | 1 | parameter must be initialized using xxx (other API) |
| BN_bn2bin | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_find | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_free | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_pop_free | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_push | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_reserve | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_set_cmp_func | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_unshift | 1 | parameter must be initialized using xxx (other API) |
| OPENSSL_sk_zero | 1 | parameter must be initialized using xxx (other API) |
| SSL_CTX_set_cipher_list | 1 | parameter must be initialized using xxx (other API) |
| SSL_shutdown | 1 | parameter must be initialized using xxx (other API) |
| EVP_PKEY_CTX_free | 1 | parameter must be initialized using xxx (other API) |
| SSL_want | 1 | parameter must be initialized using xxx (other API) |
| OSSL_PROVIDER_load | 1 | parameter must be initialized using xxx (other API) |
| SSL_CTX_get_cert_store | 1 | parameter must be initialized using xxx (other API) |
| SSL_CTX_use_certificate_chain_file | 1 | parameter must be initialized using xxx (other API) |
| EVP_DigestInit | 1 | parameter must be initialized using xxx (other API) |
| SHA384_Final | 2 | parameter must be initialized using xxx (other API) |
| X509_NAME_oneline | 1 | parameter must be initialized using xxx (other API) |
| BN_CTX_free | 1 | parameter must be initialized using xxx (other API) |
| CRYPTO_free | 1 | parameter must be initialized using xxx (other API) |
| SHA224_Init | 1 | parameter must be initialized using xxx (other API) |
| X509_get0_notAfter | 1 | parameter must be initialized using xxx (other API) |
| SSL_is_init_finished | 1 | parameter must be initialized using xxx (other API) |
| X509_STORE_CTX_set_error | 1 | parameter must be initialized using xxx (other API) |
| EVP_CIPHER_get0_name | 1 | parameter must be initialized using xxx (other API) |
| SSL_CTX_set_options | 1 | parameter must be initialized using xxx (other API) |
| CRYPTO_free | 1 | Do not call this API multiple times for the same parameter. |
| PKCS12_free | 1 | Do not call this API multiple times for the same parameter. |
| EVP_CIPHER_free | 1 | Do not call this API multiple times for the same parameter. |
| EVP_EncryptInit_ex | 4 | Parameter 4: Do not provide an IV with a size different from the cipher's block size. |
| SSL_CTX_use_certificate_chain_file | 2 | Parameter 2: The caller MUST NOT overwrite the file parameter with a buffer larger than the allowed maximum length to avoid buffer overflows |
| EVP_DigestSignUpdate | 2 | Parameter 2: Avoid passing a size larger than the actual size of the data being passed to the `data` parameter |
| SHA384_Update | 3 | Parameter 3: The caller MUST NOT set the value of `len` to a very large value such as `INT_MAX`. The length should be reasonable and within the actual length of the data |
| DES_ede3_cbc_encrypt | 3 | Parameter 3: The caller MUST NOT set the length parameter to a value greater than the maximum allowed |
| RAND_bytes | 2 | "Parameter 2 (num): Avoid configuring the buffer size larger than the size allocated in memory. |
